### PR TITLE
Add Windows Support for DeepSpeed

### LINKED ISSUE #6651: Enable Deepspeed on Windows
## 🚀 Feature
Currently, `deepspeed` is explicitly disabled on Windows systems:
```
_DEEPSPEED_AVAILABLE = not _IS_WINDOWS and _module_available('deepspeed')
```
I checked by commenting out the `not _IS_WINDOWS` part that `deepspeed` indeed does not currently work on Windows, the following error is emitted:

```
    595         elif backend == Backend.NCCL:
    596             if not is_nccl_available():
--> 597                 raise RuntimeError("Distributed package doesn't have NCCL "
    598                                    "built in")
    599             pg = ProcessGroupNCCL(
RuntimeError: Distributed package doesn't have NCCL built in
```

### Motivation

It would be awesome to get the speed and memory benefits of `deepspeed` on Windows machines too!

### Pitch

My understanding is that enabling this requires using our own implementation of parallelism via MPI rather than Pytorch distributed, which doesn't support Windows. Not sure how involved this is.

### Alternatives

N/A

### Additional context

Discussion with @SeanNaren about it here: https://pytorch-lightning.slack.com/archives/CRBLFHY79/p1616436343124400

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #6651.

To test this, I think I'll have to mock a bunch of things and ensure initialization works on Windows.

Fixes #<issue_number>

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by kaushikb11** [comment]
@SeanNaren How are you testing it on Windows?

**Comment by SeanNaren** [comment]
@kaushikb11 I've added a weak test to ensure that the function runs correctly when the platform is Windows, however actually testing this will be extremely difficult. I think we'll rely on user testing + deepspeed to ensure correct communication paths!

File: tests/plugins/test_deepspeed_plugin.py [hunk 1]
```diff
@@ -694,6 +695,26 @@ def test_deepspeed_multigpu_test(tmpdir, deepspeed_config):
...
+    """
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        plugins=[DeepSpeedPlugin(stage=3)],
+    )
+    plugin = trainer.training_type_plugin
+    assert isinstance(plugin, DeepSpeedPlugin)
+    plugin.init_ddp_connection()
+    mock_deepspeed_distributed.assert_called()
+    assert len(os.environ) == 0
```
Comment (carmocca) [inline tests/plugins/test_deepspeed_plugin.py:715]: what are you trying to test with this line?

File: tests/plugins/test_deepspeed_plugin.py [hunk 2]
```diff
@@ -694,6 +695,26 @@ def test_deepspeed_multigpu_test(tmpdir, deepspeed_config):
     trainer.test(model)
 
 
+@RunIf(deepspeed=True)
+@mock.patch('platform.system', return_value='Windows')
+@mock.patch('deepspeed.init_distributed', autospec=True)
+@mock.patch.dict(os.environ, {}, clear=True)
```
Comment (carmocca) [inline tests/plugins/test_deepspeed_plugin.py:701]: we have a conftest fixture for this already 

https://github.com/PyTorchLightning/pytorch-lightning/blob/2d682fa4618c8c70535c55c20d51e73fc367c948/tests/conftest.py#L35-L42

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 3]
```diff
@@ -340,6 +341,30 @@ def setup_distributed(self):
...
+        )
+
+    def _set_node_environment_variables(
+        self, global_rank: Optional[int] = None, world_size: Optional[int] = None
+    ) -> None:
+        os.environ["MASTER_ADDR"] = self.cluster_environment.master_address()
+        os.environ["MASTER_PORT"] = str(self.cluster_environment.master_port())
+        os.environ["RANK"] = str(global_rank)
+        os.environ["WORLD_SIZE"] = str(world_size)
+        os.environ["LOCAL_RANK"] = str(self.local_rank)
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:366]: noob q: isn't setting these responsibilities of the cluster environment?

Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:366]: I think the cluster environments only provide the variables, assuming that others have set it for the environment we're in. In our case we're passing the set variables into a format for DeepSpeed to setup internally

File: tests/plugins/test_deepspeed_plugin.py (hunk 1 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py]: I've expanded the test, please re-review!

File: tests/plugins/test_deepspeed_plugin.py (hunk 2 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py]: Thanks, I've changed the test now to not require this

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Pull request merged**